### DISCUSSION TITLE
`executemany()` individual error handling

### DISCUSSION THREAD

**Discussion by KSmanis** [psycopg/psycopg#726] [General]
Is it possible to handle individual errors in a `executemany()` batch?

For instance, assuming a `customers` table:
```sql
CREATE TABLE customers (
    id VARCHAR(20) PRIMARY KEY
);
```
is it possible to handle a primary key collision like this:
```python
with psycopg.connect() as conn, conn.cursor() as cur:
    cur.executemany("INSERT INTO customers(id) VALUES (%s)", [["1"], ["1"]])
```

**Comment by dvarrazzo** [comment]
I don't think so

**Reply by KSmanis** [reply]
What would be a more performant alternative to `INSERT`ing rows in a loop that allows for DB error handling on a per-statement basis?

**Reply by dlax** [reply]
You can use the [pipeline mode](https://www.psycopg.org/psycopg3/docs/advanced/pipeline.html) (used under the hood by `executemany()` if available libpq supports it):
```python
with conn.pipeline():
   for args in ...:
       conn.execute("INSERT INTO customers(id) VALUES (%s)", args)
```
